core/src/main/scala-3/sttp/tapir/macros/ValidatorMacros.scala

```diff
@@ -1,12 +1,49 @@
 package sttp.tapir.macros
 
 import sttp.tapir.Validator
+import sttp.tapir.SchemaType
+
+import scala.compiletime
+
+import scala.quoted.*
 
 trait ValidatorMacros {
 
   /** Creates an enum validator where all subtypes of the sealed hierarchy `T` are `object`s.
     * This enumeration will only be used for documentation, as a value outside of the allowed values will not be
     * decoded in the first place (the decoder has no other option than to fail).
     */
-  def derivedEnumeration[T]: Validator.Enumeration[T] = ??? // TODO
+  inline def derivedEnumeration[T]: Validator.Enumeration[T] = ${ValidatorMacros.derivedEnumerationImpl[T]}
 }
+
+object ValidatorMacros {
+  def derivedEnumerationImpl[T: Type](using q: Quotes): Expr[Validator.Enumeration[T]] = {
+    import quotes.reflect.*
+
+    val tpe = TypeRepr.of[T]
+    val symbol = tpe.typeSymbol
+
+    if (!symbol.isClassDef || !(symbol.flags is Flags.Sealed)) {
+      report.throwError("Can only enumerate values of a sealed trait, class or enum.")
+    } else {
+      val children = symbol.children.toList.sortBy(_.name)
+      
+      if (!children.forall(_.flags is Flags.Module) && !(symbol.flags is Flags.Enum)) {
+        report.throwError("All children must be objects or enum cases.")
```

Comment:
enum cases can also have parameters. So we should reject e.g.:

```
enum Color:
   case Red(x: String)
```

but accept:

```
enum Color:
   case Red
```

but maybe this is already covered? I think we'll need a scala3-specific test here

Reply:
added test and improved error message for this scenario